client/state/data-layer/http-data.ts

```diff
@@ -0,0 +1,345 @@
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+
+/**
+ * Types
+ */
+import { Lazy, TimestampMS, TimerHandle } from 'types';
+
+enum DataState {
+	Failure = 'failure',
+	Pending = 'pending',
+	Success = 'success',
+	Uninitialized = 'uninitialized',
+}
+
+interface ResourceData {
+	state: DataState;
+	data: any;
+	error: any;
+	lastUpdated: TimestampMS;
+	pendingSince: TimestampMS | undefined;
+}
+
+type Resource =
+	| ResourceData & {
+			state: DataState.Uninitialized;
+			data: undefined;
+			error: undefined;
+			pendingSince: undefined;
+	  }
+	| ResourceData & { state: DataState.Pending; error: undefined }
+	| ResourceData & { state: DataState.Failure; pendingSince: undefined }
+	| ResourceData & { state: DataState.Success; error: undefined; pendingSince: undefined };
+
+type DataId = string;
+
+export const httpData = new Map< DataId, Resource >();
+export const listeners = new Set< () => void >();
+
+export const empty: Readonly< Resource > = Object.freeze( {
+	state: DataState.Uninitialized,
+	data: undefined,
+	error: undefined,
+	lastUpdated: -Infinity,
+	pendingSince: undefined,
+} );
+
+export const getHttpData = ( id: DataId ) => httpData.get( id ) || empty;
+
+export const subscribe = ( f: () => void ): ( () => void ) => {
+	listeners.add( f );
+
+	return () => void listeners.delete( f );
+};
+
+export const updateData = ( id: DataId, state: DataState, data: unknown ) => {
+	const lastUpdated: TimestampMS = Date.now();
+	const item = httpData.get( id ) || empty;
+
+	// We could have left out the keys for
+	// the previous properties if they didn't
+	// exist but I wanted to make sure we can
+	// get our hidden classes to optimize here.
+	switch ( state ) {
+		case DataState.Failure:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: data,
+				lastUpdated: item.lastUpdated,
+				pendingSince: undefined,
+			} );
+
+		case DataState.Pending:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: undefined,
+				lastUpdated: item.lastUpdated,
+				pendingSince: lastUpdated,
+			} );
+
+		case DataState.Success:
+			return httpData.set( id, {
+				state,
+				data,
+				error: undefined,
+				lastUpdated,
+				pendingSince: undefined,
+			} );
+	}
+};
+
+export const update = ( id: DataId, state: DataState, data?: unknown ) => {
+	const updated = updateData( id, state, data );
+
+	listeners.forEach( f => f() );
+
+	return updated;
+};
+
+const fetch = ( action: HttpDataAction ) => {
+	update( action.id, DataState.Pending );
+
+	return [
+		{
+			...action.fetch,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
+		},
+		{ type: HTTP_DATA_TICK },
+	];
+};
+
+const onError = ( action: HttpDataAction, error: any ) => {
+	update( action.id, DataState.Failure, error );
+
+	return { type: HTTP_DATA_TICK };
+};
+
+type SuccessfulParse = [undefined, ReturnType< ResponseParser >];
+type FailedParse = [Error, undefined];
```

Comment:
Anything can be thrown:

```suggestion
type FailedParse = [any, undefined];
```